src/ComponentProvider.ts

```diff
@@ -1,94 +1,60 @@
-import type { AsyncAPIDocument, ChannelParameter, Message, MessageTrait, Schema } from '@asyncapi/parser';
-
+import type {
+  AsyncAPIDocument,
+  ChannelParameter,
+  Message,
+  Schema,
+} from '@asyncapi/parser';
+
+import { JSONPath } from 'jsonpath-plus';
+import _ from 'lodash';
 /**
  * This class will provide all sorts of data for optimizers.
  *
  * @private
  */
 export class ComponentProvider {
+  messagePaths = ['$.channels.*.*.message', '$.components.messages.*'];
+  schemaPaths = [
+    '$.channels.*.*.message.traits[*]..[?(@.type)]',
+    '$.channels.*.*.message.headers',
+    '$.channels.*.*.message.headers..[?(@.type)]',
+    '$.channels.*.*.message.payload',
+    '$.channels.*.*.message.payload..[?(@.type)]',
+    '$.channels.*.parameters.*.schema[?(@.type)]',
+    '$.channels.*.parameters.*.schema..[?(@.type)]',
+    '$.components.schemas..[?(@.type)]',
```

Comment:
Please have in mind that you don't need to define the `type` field in the schema, so you will skip some schemas. I can accept it for now, but we should fix that.

Reply:
I was thinking about it and I couldn't find a perfect way to define a `schema worth optimizing`. If we set the bar too low like `all schemas should be optimized if it is an object, then the library will extract an unnecessary amount of schemas.
like `{type: string}` should be ignored since by optimizing it, we make it complicated and hard to understand.

if we set the bar too high then it will skip some schemas.

what is the perfect balance here?
the best that I could come up with is:
```
a schema should be optimized if: 
* it has more than one non-object or non-array field
or
* it has at least one object or array field. 
```

what do you think?

Reply:
For that delay 😅 

> it has at least one object or array field.

Yeah, it's good atm :) We should remember about that "restriction". Could you create issue for that?

Reply:
@magicmatatjahu  Sorry, I don't get it, should I leave it as it is or implement the restriction that I have proposed? or remove that having a `type` field restriction entirely?

Reply:
as you suggested :)

Reply:
but, that PR is a refactor, so we won't release it. I can accept it and you will work on it in the next PR, what do you think?

Reply:
@magicmatatjahu that would be great. can you approve this PR then?